pkg/resource/target_reconciler.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+)
+
+const (
+	targetControllerName   = "kubernetestarget.crossplane.io"
+	targetReconcileTimeout = 1 * time.Minute
+
+	errGetTarget                 = "unable to get Target"
+	errMissingSecretRef          = "no secretRef specified for Target with a clusterRef"
+	errManagedResourceIsNotBound = "managed resource in Target clusterRef is unbound"
+	errUpdateTarget              = "unable to update Target"
+)
+
+// A TargetKind contains the type metadata for a kind of target resource.
+type TargetKind schema.GroupVersionKind
+
+// A TargetReconciler reconciles targets by propagating the secret of the
+// referenced managed resource.
+type TargetReconciler struct {
+	client     client.Client
+	newTarget  func() Target
+	newManaged func() Managed
+
+	propagator ManagedConnectionPropagator
+}
+
+// NewTargetReconciler returns a Reconciler that reconciles
+// KubernetesTarget by setting their ConnectionSecretRef to that of the
+// referenced Kubernetes cluster implementation.
+func NewTargetReconciler(m manager.Manager, of TargetKind, with ManagedKind) *TargetReconciler {
+	nt := func() Target { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Target) }
+	nr := func() Managed { return MustCreateObject(schema.GroupVersionKind(with), m.GetScheme()).(Managed) }
+
+	// Panic early if we've been asked to reconcile a target or resource kind
+	// that has not been registered with our controller manager's scheme.
+	_, _ = nt(), nr()
+
+	r := &TargetReconciler{
+		client:     m.GetClient(),
+		newTarget:  nt,
+		newManaged: nr,
+		propagator: NewAPIManagedConnectionPropagator(m.GetClient(), m.GetScheme()),
+	}
+
+	return r
+}
+
+// Reconcile a target with a concrete managed resource.
+func (r *TargetReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("Reconciling", "controller", targetControllerName, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), targetReconcileTimeout)
+	defer cancel()
+
+	target := r.newTarget()
+	if err := r.client.Get(ctx, req.NamespacedName, target); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetTarget)
+	}
+
+	if target.GetWriteConnectionSecretToReference() == nil {
+		// If the ConnectionSecretRef is not set on this Target, we will not
+		// know where to propagate the secret. We do not explicitly requeue
+		// because we have a watch on updates to the KubernetesTarget.
+		target.SetConditions(v1alpha1.SecretPropagatedError(errors.New(errMissingSecretRef)))
+		return reconcile.Result{Requeue: false}, errors.Wrap(r.client.Status().Update(ctx, target), errUpdateTarget)
+	}
+
+	if meta.WasDeleted(target) {
```

Comment:
Does the secret created by this reconciler have the owner reference of this object for cleanup after it's deleted? I checked the propagator implementation but couldn't catch it.

Reply:
Yep! The propagator calls `LocalConnectionSecretFor` on the `Target` which sets owner references. Importantly, we also plan to add an owner reference if a `Target` references an existing `Secret` directly (not part of this PR) such that if a `Secret` is manually created in a `Namespace` and a `KubernetesTarget` is created to make it a `KubernetesApplication` scheduling target, it will automatically be cleaned up by garbage collection when the `Target` is deleted. In the case that a `KubernetesTarget` is created that references an existing `Secret` that was claimed in the `Namespace`, we plan to add an owner reference to that `Secret` such that it is not garbage collected until both the `KubernetesCluster` claim and the `KubernetesTarget` are deleted. This would mean that a `KubernetesTarget` could reference a `Secret` that will not allow for successful connection to a k8s cluster (in the case that the `KubernetesCluster` claim was deleted and `Reclaim: Delete`), but this seems fine due to the fact that `KubernetesTarget` makes no claims (apologize for using an overloaded term) about making a successful connection, just rather that a `Secret` with connection info exists in the `Namespace`.

https://github.com/crossplaneio/crossplane-runtime/blob/d5c9dedd2a61108120fb5b3bba32e5a8d43ffeee/pkg/resource/resource.go#L45